architecture-decision-records/0022-cloudfront-upload-proxy.md

```diff
@@ -0,0 +1,164 @@
+# 22. Proxy S3 uploads through CloudFront
+
+**Date:** 2020-09-17
+
+## Context
+
+When a user uploads a file to TDR, the file is currently sent from the JavaScript running in the user's browser straight
+to the AWS S3 upload bucket. This means that there are several HTTP requests from the user's browser to AWS domains:
+
+- `cognito-identity.eu-west-2.amazonaws.com`: Exchange a TDR access token for a Cognito token
+- `sts.amazonaws.com`: Exchange the Cognito token for a long-lived STS token
+- `tdr-upload-files-dirty-<environment>.s3.eu-west-2.amazonaws.com`: Upload the files
+
+The sequence of requests looks like this (simplified slightly - CORS requests and multiple requests to the auth server
+and S3 are omitted for clarity):
+
+![Sequence diagram showing current upload process](images/upload-sequence-diagram-cognito.png)
```

Comment:
It's a minor thing which doesn't take away from what you're saying but the diagram says that we exchange the Cognito token for an STS token. We actually call STS to assume a role using the Cognito token and that returns AWS credentials which we then pass to the S3 SDK. So instead of `exchange Cognito for long-lived STS token` it's more `use Cognito token to assume role with STS which returns AWS credentials` and then `check token and S3 key against IAM policy` is `check the S3 key against the IAM policy of the assumed role attached to the credentials`
Or something along those lines anyway. 
This is [the relevant STS call](https://github.com/nationalarchives/tdr-transfer-frontend/blob/master/npm/src/auth/index.ts#L96) which shows us calling `assumeRoleWithWebIdentity`

Reply:
Ahh, I'd forgotten the details of that call, thanks. I've reworded mentions of "STS token" to make it clear that STS gives us temporary AWS credentials with a session token. I've updated the first sequence diagram too. Does it make sense?

Reply:
Yep, looks good, thanks.